rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,475 @@
+
+```
+.foo {
+  --a: green;
+  --b: yellow;
+}
+
+vs
+
+.foo:enabled:not(:checked):not(:indeterminate)' > svg
+```
+
+### 👍
+
+### 👎
+
+Composition approach would be difficult to adjust at runtime because css vars are already set on the root, and new styles would either need to replace the old one, be more specific, or not use makeResetStyles so user could override individual css vars
+
+### 🤔
+
+## Option D: Build time component tokens
+
+Each component would define a hierarchy component-level token constants.
+Component tokens would reference alias tokens.
+The hierarchy would have levels for slot and state variations.
+The component style hook would then use the component token constants rather than alias token constants.
+The component's tokens could be substituted at build time for a mapping to other tokens (CSS var references)
+
+```ts
+// Button.tokens.ts (limited example)
+export const buttonTokens = {
+  root: {
+    backgroundColor: tokens.colorNeutralBackground1,
+    hover: {
+      backgroundColor: tokens.colorNeutralBackground1Hover,
+    }
+    active: {
+      backgroundColor: tokens.colorNeutralBackground1Pressed
+    },
+    outline: {
+      backgroundColor: tokens.colorTransparentBackground,
+      hover: {
+        backgroundColor: tokens.colorTransparentBackgroundHover
+      },
+      active: {
+        backgroundColor: tokens.colorTransparentBackgroundPressed
+      }
+    }
+  },
+  icon: {
+    small: {
+      spacing: tokens.spacingHorizontalXS
+    },
+    medium: {
+      spacing: tokens.spacingHorizontalSNudge
+    },
+    large: {
+      spacing: tokens.spacingHorizontalSNudge
+    }
+  }
+}
+```
+
+### 👍
+
+- Codifies the mapping of component part style to alias token.
+- Build time substitution allows for continued build time style optimizations
+- Runtime CSS vars are unchanged and works with existing themes
+- Provides a similar experience to custom styles and theme component styles from v8, but mapped to slots.
+
+### 👎
+
+- Exports yet another set of constants for the bundle
+- Adds some complexity to the style hooks
+
+### 🤔
+
+- Would the hierarchy of component tokens be a contract?
+
+## Decision making
+
+//TODO @Geoff
+
+## Appendix: Partner scenarios
+
+### MADS: Centro use case
+
+from Mason Tejera
+
+Our primary customer for applying component specific styles via the ThemeProvider has been Centro. The Core mission of Centro is to allow engineering team to build experiences (UX + Backend) one time and ship them to any hosting application(s). They are a horizontal engineering team that services Commercial and Consumer experiences. They surface many of their experiences across multiple apps & platforms and require them to look native to the end user. This sort of functionality saves Centro and their consumers a LOT of engineering time. A couple specific examples of how they use it:
```

Comment:
It would be great if we could hear more from @mltejera on this one.
One of the goals of v9 vs v8/0 was to allow experiences reusable across apps without requiring any style tweaks by implementing Fluent Design + tokens. Here I read the situation is worse in v9.
If that's the case, we need to learn from that and also share the learning with design and wider audience.

This might be a temporary problem which will be solved by applications moving to the Fluent design - then it is fine and we must help these partners as much as possible to make their migration path smooth.

This might also be a permanent problem - those are applications which do not adhere to Fluent design - then we should question whether they should stick to Fluent UI or use something else (like we can provide them with unstyled components).

Reply:
Sure Miro - I do think this is a permanent problem/opportunity inherent in a company with as much breadth as Microsoft.

A goal to 'minimize style changes' is very limiting at a company with our a diverse a user base as Microsoft. The better goal should be that Fluent can be used by kindergartners in the EDU space, IT Pro's, gamers, and on the go in every day life. 

Our users are on EVERY modern computing platform - iOS, Android, MacOS, Windows, Web and probably more we don't know of today. An app should look and feel like a hybrid of Fluent and iOS when it's on a phone, and a hybrid of Fluent and Windows when on a desktop browser.

A token based system assumes we can predict every place and way consumers will want to customize.  I know from 3+ years of working on a Fluent extension library that this is not a safe assumption. We (Fluent) needs to be more humble. I've seen enough diversity in the IT Pro space alone that Fluent needs a richer customization story beyond tokens.

We have research that shows most internal Fluent consumers are using Fluent through an extension library. We should always assume the consuming team knows more about their users than we ever will, and that they are highly skilled professionals that are on the same Microsoft Team.

I've met apps teams who've looked at v9 and are deciding to stick with v8 because v9 is too opinionated. Worse, some are deciding to move to a 3rd party solution because v8 support has stopped. We're getting 0% coherence because we're unable to budge from 98% to 90%. We have to assume those teams have solid reasons for needing that extra 8%.

I think *every* styling opinion in the library should be efficient and scalable to disagree with. We should provide the mechanisms to do so instead of telling each team create their own customization layer. This maximizes code re-use across the company and will ultimately improve coherence.